dashboard/app/controllers/api_controller.rb

```diff
@@ -184,8 +184,22 @@ def section_progress
 
     script_levels = script.script_levels.where(bonus: nil)
 
+    # Clients are seeing requests time out for large sections as we attempt to
+    # send back all of this data. Allow them to instead request paginated data
+    if params[:page] && params[:per]
+      paged_students = section.students.page(params[:page]).per(params[:per])
+      # As designed, if there are 50 students, the client will ask for both
+      # page 1 and page 2, even though page 2 is out of range. However, it should
+      # never ask for page 3
+      if params[:page].to_i > paged_students.total_pages + 1
+        return head :range_not_satisfiable
+      end
```

Comment:
Instead of having the client fail if it starts requesting too many pages, I have the server fail if it's requesting a page that definitely out of range (which should also stop the client from asking for more pages if things are working right).